src/java.rs

```diff
@@ -16,61 +16,46 @@ fn java_base_types(encoded_ty: char) -> Option<&'static str> {
 }
 
 fn byte_code_type_to_java_type(byte_code_type: &str, mapper: &ProguardMapper) -> Option<String> {
-    let mut chrs = byte_code_type.chars();
-    let token = chrs.next()?;
-    if token == 'L' {
-        // invalid signature
-        let l = chrs.clone().last();
-        if l.is_none() || l.unwrap() != ';' {
-            return None;
-        }
-        chrs.next_back(); // remove final `;`
-        let obfuscated = chrs.collect::<String>().replace('/', ".");
+    let mut chrs = byte_code_type.char_indices();
+    //let (idx, token) = chrs.next()?;
+    let mut suffix = String::new();
+    while let Some((idx, token)) = chrs.next() {
+        if token == 'L' {
+            // expect and remove final `;`
+            if chrs.next_back()?.1 != ';' {
+                return None;
+            }
+            let obfuscated = byte_code_type
+                .get(idx + 1..byte_code_type.len() - 1)?
+                .replace('/', ".");
 
-        if let Some(mapped) = mapper.remap_class(&obfuscated) {
-            return Some(mapped.to_string());
-        }
+            if let Some(mapped) = mapper.remap_class(&obfuscated) {
+                return Some(format!("{}{}", mapped, suffix));
+            }
 
-        return Some(obfuscated);
-    } else if token == '[' {
-        let type_sig = chrs.clone().collect::<String>();
-        if !type_sig.is_empty() {
-            return Some(format!(
-                "{}[]",
-                byte_code_type_to_java_type(chrs.collect::<String>().as_str(), mapper)
-                    .unwrap_or_default()
-            ));
+            return Some(format!("{}{}", obfuscated, suffix));
+        } else if token == '[' {
+            suffix.push_str("[]");
+            continue;
+        } else if let Some(ty) = java_base_types(token) {
+            return Some(format!("{}{}", ty, suffix));
         }
-    } else if let Some(ty) = java_base_types(token) {
-        return Some(ty.to_string());
     }
-    Some(byte_code_type.to_string())
+    None
 }
 
 // parse_obfuscated_bytecode_signature will parse an obfuscated signatures into parameter
 // and return types that can be then deobfuscated
 fn parse_obfuscated_bytecode_signature(signature: &str) -> Option<(Vec<String>, String)> {
-    let mut chrs = signature.chars();
-
-    let token = chrs.next();
-    if token.unwrap_or_default() != '(' {
-        return None;
-    }
-
-    let sig = chrs.collect::<String>();
-    let split_sign = sig.rsplitn(2, ')').collect::<Vec<&str>>();
-    if split_sign.len() != 2 {
-        return None;
-    }
+    let signature = signature.strip_prefix('(')?;
 
-    let return_type = split_sign[0];
-    let parameter_types = split_sign[1];
+    let (parameter_types, return_type) = signature.rsplit_once(')')?;
     if return_type.is_empty() {
         return None;
     }
 
     let mut types: Vec<String> = Vec::new();
-    let mut tmp_buf: Vec<char> = Vec::new();
+    let mut tmp_buf: String = String::new();
```

Comment:
ohhhh, now I finally understand what this function is even doing. It is just splitting the string into different arguments.
So essentially you don’t need a `tmp_buf` at all, and your `types` can just be a `Vec<&str>`.

Because all you are doing is collect chars into that buffer until the end of one parameter is reached.